### PR TITLE
Build HeadRelayer / UpkeepExecuter as Keeper Delegate Service

### PR DESCRIPTION
**Pull request by RyanRHall**
https://www.pivotaltracker.com/story/show/177238754

### CODE REVIEW

File: core/services/head_relayer.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	hr.callbacks[id] = callback
+	return func() {
+		hr.mutex.Lock()
+		defer hr.mutex.Unlock()
+		delete(hr.callbacks, id)
+	}
+}
+
+func (hr *HeadRelayer) run() {
+	hr.wgDone.Add(1)
```
Comment (samsondav) [inline core/services/head_relayer.go:95]: This should be pulled out of the goroutine, otherwise its racey

File: core/services/head_relayer.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+		}
+	}
+}
+
+// DEV: the head relayer makes no promises about head delivery! Subscribing
+// Jobs should expect to the relayer to skip heads if there is a large number of listeners
+// and all callbacks cannot be completed in the allotted time.
+func (hr *HeadRelayer) executeCallbacks() {
+	hr.mutex.RLock()
+	defer hr.mutex.RUnlock()
```
Comment (samsondav) [inline core/services/head_relayer.go:112]: This keeps the mutex locked for a long time. Maybe we could lock just long enough to copy the slice and then unlock and use our copied version?

File: core/services/keeper/upkeep_executer.go [hunk 3]
```diff
@@ -71,19 +81,8 @@ func (executor *UpkeepExecutor) Close() error {
...
-	executor.blockHeight.Store(head.Number)
-	// avoid blocking if signal already in buffer
-	select {
-	case executor.chSignalRun <- struct{}{}:
-	default:
-	}
+	executor.mailbox.Deliver(head)
 }
 
 func (executor *UpkeepExecutor) run() {
```
Comment (samsondav) [inline core/services/keeper/upkeep_executer.go:88]: wg below is racey

File: core/services/keeper/upkeep_executer.go [hunk 4]
```diff
@@ -24,40 +24,50 @@ const (
...
+	orm            ORM
+	wgDone         sync.WaitGroup
 	utils.StartStopOnce
 }
 
 func (executor *UpkeepExecutor) Start() error {
 	return executor.StartOnce("UpkeepExecutor", func() error {
 		go executor.run()
+		unsubscribe := executor.headRelayer.Subscribe(executor)
+		executor.wgDone.Add(1)
```
Comment (samsondav) [inline core/services/keeper/upkeep_executer.go:65]: Probably just calling it `executor.wg` is clear enough

**Review by samsondav — CHANGES_REQUESTED** [review]
LGTM some minor comments

File: core/services/keeper/upkeep_executer.go (hunk 4 above)
Comment (RyanRHall) [reply core/services/keeper/upkeep_executer.go:65]: I like the consistency of calling it the same name across different services that implement this pattern. And using `wgDone` differentiates it from other local `wg` variables in other places in the code base.

File: core/services/head_relayer.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+		mailbox:       utils.NewMailbox(1),
+		mutex:         &sync.RWMutex{},
+		chClose:       make(chan struct{}),
+		wgDone:        sync.WaitGroup{},
+		StartStopOnce: utils.StartStopOnce{},
+	}
+}
+
+// HeadRelayer relays heads from the head tracker to subscribed jobs, it is less robust against
+// congestgion than the head tracker, and missed heads should be expected by consuming jobs
```
Comment (samsondav) [inline core/services/head_relayer.go:42]: congestion

File: core/services/head_relayer.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+		}
+	}
+}
+
+// DEV: the head relayer makes no promises about head delivery! Subscribing
+// Jobs should expect to the relayer to skip heads if there is a large number of listeners
+// and all callbacks cannot be completed in the allotted time.
+func (hr *HeadRelayer) executeCallbacks() {
+	hr.mutex.RLock()
+	callbacks := hr.copyCallbacks()
```
Comment (samsondav) [inline core/services/head_relayer.go:114]: Can you use a regular `copy` here?

**Review by samsondav — APPROVED** [review]

**Review by samsondav — DISMISSED** [review]

File: core/services/head_relayer.go [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+	for _, callback := range callbacks {
+		go func(hr HeadRelayable) {
+			start := time.Now()
+			ctx, cancel := context.WithTimeout(context.Background(), callbackTimeout)
+			defer cancel()
+			hr.OnNewLongestChain(ctx, head)
+			elapsed := time.Since(start)
+			logger.Debugw(fmt.Sprintf("HeadRelayer: finished callback in %s", elapsed), "callbackType", reflect.TypeOf(hr), "blockNumber", head.Number, "time", elapsed, "id", "head_relayer")
+			wg.Done()
```
Comment (spooktheducks) [inline core/services/head_relayer.go:134]: This has to be deferred (and moved to the first line of the goroutine), or else a panic in a callback could cause the entire `HeadRelayer` to freeze up

File: core/services/head_relayer.go [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+		}(callback)
+	}
+
+	wg.Wait()
+}
+
+func (hr *HeadRelayer) copyCallbacks() callbackSet {
```
Comment (spooktheducks) [inline core/services/head_relayer.go:141]: Purely stylistic and inconsequential, but if `callbackSet` is already its own type, you could just put this method directly on it

File: core/services/keeper/upkeep_executer.go [hunk 9]
```diff
@@ -24,40 +24,50 @@ const (
...
 func NewUpkeepExecutor(
+	job job.Job,
 	db *gorm.DB,
 	ethClient eth.Client,
+	headRelayer *services.HeadRelayer,
 ) *UpkeepExecutor {
 	return &UpkeepExecutor{
-		blockHeight:    atomic.NewInt64(0),
-		wgDone:         sync.WaitGroup{},
+		chStop:         make(chan struct{}),
```
Comment (spooktheducks) [inline core/services/keeper/upkeep_executer.go:33]: Not worth blocking the PR on, but would be a little more tidy if this line was right above the `wgDone` line

**Review by spooktheducks — CHANGES_REQUESTED** [review]

**Review by spooktheducks — APPROVED** [review]

**Pull request merged**